### PR TITLE
Make XataApiClient to use ES Proxies

### PR DESCRIPTION
**Pull request by SferaDev**

### CODE REVIEW

File: packages/client/src/api/client.ts [hunk 1]
```diff
@@ -21,1907 +18,72 @@ export interface XataApiClientOptions {
...
+      return new Proxy(this, {
+        get: (_target, namespace: keyof typeof operationsByTag) => {
+          if (operationsByTag[namespace] === undefined) {
+            return undefined;
+          }
+
+          return new Proxy(
+            {},
+            {
+              get: (_target, operation: keyof (typeof operationsByTag)[keyof typeof operationsByTag]) => {
```
Comment (eemmiillyy) [inline packages/client/src/api/client.ts:71]: Does this intercept  all operations/methods inside each API class? I guess the fetch functions that these operations/methods call are going to be intercepted?

File: packages/client/src/api/client.ts [hunk 2]
```diff
@@ -21,1907 +18,72 @@ export interface XataApiClientOptions {
...
-    database: Schemas.DBName;
-  }): Promise<void> {
-    return operationsByTag.databases.deleteDatabaseGithubSettings({
-      pathParams: { workspaceId: workspace, dbName: database },
-      ...this.extraProps
-    });
-  }
+      });
+    }
+  } as unknown as { new (options?: XataApiClientOptions): XataApiProxy };
```
Comment (eemmiillyy) [inline packages/client/src/api/client.ts:87]: Is the `new` part necessary? This syntax is confusing/new to me

File: packages/client/src/api/client.ts [hunk 3]
```diff
@@ -21,1907 +18,72 @@ export interface XataApiClientOptions {
...
-      | {
-          table: string;
-          filter?: Schemas.FilterExpression;
-          target?: Schemas.TargetExpression;
-          boosters?: Schemas.BoosterExpression[];
+type UserProps = {
+  headers?: Record<string, unknown>;
+};
+
+type XataApiProxy = {
```
[the remark this answers is not in the document]
Comment (SferaDev) [reply packages/client/src/api/client.ts:25]: Yes, we need to rewrite the return type as the proxy won't infer the different properties/namespaces we will expose.

File: packages/client/src/api/client.ts [hunk 4]
```diff
@@ -21,1907 +18,72 @@ export interface XataApiClientOptions {
...
+        apiKey,
+        trace,
+        clientName: options.clientName,
+        xataAgentExtra: options.xataAgentExtra,
+        clientID
+      };
+
+      return new Proxy(this, {
+        get: (_target, namespace: keyof typeof operationsByTag) => {
+          if (operationsByTag[namespace] === undefined) {
```
[the remark this answers is not in the document]
Comment (SferaDev) [reply packages/client/src/api/client.ts:64]: Exactly, this is a catch for runtime JS where a user tries to access a namespace that doesn't exist

File: packages/client/src/api/client.ts (hunk 1 above)
Comment (SferaDev) [reply packages/client/src/api/client.ts:71]: The `get` here does not refer to HTTP `GET` operation. The `get` in a `Proxy` is trying to access a property/field in an object like `foo.bar`. In this particular case, in the class it's intercepting trying to read any member of it.

File: packages/client/src/api/client.ts (hunk 2 above)
Comment (SferaDev) [reply packages/client/src/api/client.ts:87]: Haha, yeah! This is how `class` types are defined where the `new` clause refers to the constructor of a class. As a fun fact, classes are syntactic sugar for objects with a constructor that is injected in the new property.

File: packages/client/src/api/client.ts [hunk 5]
```diff
@@ -1,14 +1,11 @@
 import { defaultTrace, TraceFunction } from '../schema/tracing';
 import { getAPIKey } from '../util/environment';
 import { FetchImpl, getFetchImplementation } from '../util/fetch';
+import { FlattenObject, RequiredKeys } from '../util/types';
```
Comment (SferaDev) [inline packages/client/src/api/client.ts:4]:
```suggestion
import { RequiredKeys } from '../util/types';
```

File: test/integration/smoke.test.ts [hunk 6]
```diff
@@ -88,7 +88,7 @@ describe('API Client Integration Tests', () => {
         tableName: 'table'
       },
       body: {
-        record: { email: 'example@foo.bar' }
+        email: 'example@foo.bar'
```
Comment (eemmiillyy) [inline test/integration/smoke.test.ts:91]: We may have had a bug here and didn't notice. I don't think this part of the API was changed

Comment (SferaDev) [reply test/integration/smoke.test.ts:91]: I just added this test on Friday, weird I didn't catch it

**Review by eemmiillyy — APPROVED** [review]

**Pull request merged**